src/ripple/shamap/impl/TaggedPointer.ipp

```diff
@@ -33,10 +33,34 @@ namespace {
 // Given n children, an array of size `*std::lower_bound(boundaries.begin(),
 // boundaries.end(), n);` is used to store the children. Note that the last
 // element must be the number of children in a dense array.
+//
+//  We base our boundaries on this sampling of inner nodes from
+//  ledger  83,592,201:
+//
+//  +----------+-----------+
+//  | Children |     Count |
+//  +----------+-----------+
+//  |        1 |   578,512 |
+//  |        2 | 2,840,810 | <-------
+//  |        3 |   624,837 |
+//  |        4 |   120,458 | <-------
+//  |        5 |    45,802 |
+//  |        6 |    73,588 |
+//  |        7 |   129,694 |
+//  |        8 |   184,122 |
+//  |        9 |   208,545 |
+//  |       10 |   187,800 | <-------
```

Comment:
@mtrippled @seelabs

The sampling data from ledger 83,592,201 suggests a potential optimization opportunity by adjusting boundaries to cater to common child counts in inner nodes. This could lead to memory allocation improvements.

Thoughts?

Reply:
I can confirm that given these counts, these are the optimal boundaries, when limited to exactly four, that minimize the allocated size. I brute-forced the computation in a [spreadsheet](https://docs.google.com/spreadsheets/d/13PUUJrpNc-gvbQwTNoxR-keMVGRajpEu6EsghvLGPy0/edit?usp=sharing). A closed form solution wasn't coming to me, but I'm interested to know if there is one.

Reply:
@thejohnfreeman 

Thanks for looking. 

p.s. feel free to push to this branch